junit5/src/main/java/cz/xtf/junit5/extensions/OpenShiftRecorderService.java

```diff
@@ -0,0 +1,527 @@
+        } catch (CloneNotSupportedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private ExtensionContext.Store getClassStore(ExtensionContext extensionContext) {
+        return extensionContext.getStore(ExtensionContext.Namespace.create(extensionContext.getRequiredTestClass()));
+    }
+
+    private ExtensionContext.Store getMethodStore(ExtensionContext extensionContext) {
+        return extensionContext
+                .getStore(ExtensionContext.Namespace.create(extensionContext.getRequiredTestClass(),
+                        extensionContext.getTestClass()));
+    }
+
+    private <E extends HasMetadata> ResourcesFilterBuilder<E> getFilter(ExtensionContext context, String key) {
+        OpenShiftRecorder classOpenShiftRecorder = AnnotationSupport
+                .findAnnotation(context.getRequiredTestClass(), OpenShiftRecorder.class).orElse(null);
+        OpenShiftRecorder methodOpenShiftRecorder = AnnotationSupport
+                .findAnnotation(context.getElement(), OpenShiftRecorder.class).orElse(null);
+        OpenShiftRecorder openShiftRecorder = methodOpenShiftRecorder != null
+                ? methodOpenShiftRecorder
+                : classOpenShiftRecorder;
+
+        // annotation (openShiftRecorder is not null) or global include in META-INF.services - null
+        String[] resourceNames = openShiftRecorder != null ? openShiftRecorder.resourceNames() : null;
+
+        ResourcesFilterBuilder<E> filter = context.getTestMethod().isPresent()
+                ? (ResourcesFilterBuilder<E>) getMethodStore(context).get(key, ResourcesFilterBuilder.class)
+                : (ResourcesFilterBuilder<E>) getClassStore(context).get(key, ResourcesFilterBuilder.class);
+
+        // OpenShiftRecorder (OpenShiftRecorderHandler) may be used in following ways
+        // (1) A test is annotated by @OpenShiftRecorder - OpenShiftRecorder#resourcesNames is set to default value -
+        //     an array of one empty string.
+        // (2) A test is annotated by @OpenShiftRecorder(string [string]*) -
+        //     OpenShiftRecorder#resourcesNames is an array of strings. null value is not allowed in an annotation.
+        // (3) OpenShiftRecorderHandler is defined via SPI in resources/META-INF/services
+        //     the annotation is not found for the test class and resourceNames is null
+        if (resourceNames != null // exclude (3)
+                && !(resourceNames.length == 1 && resourceNames[0].equals("")) // exclude (1)
+        ) {
+            // option (2) list of resource names is present - filter by names
+            filter.filterByResourceNames();
+            filter.setResourceNames(resourceNames);
+        } else {
+            filter.filterByLastSeenResources();
+        }
+        return filter;
+    }
+
+    private void saveStatefulsets(ExtensionContext context, ResourcesFilterBuilder<StatefulSet> masterFilter)
```

Comment:
we have a chance now to make it extendable, what about changing methods visibility to `protected`?

Reply:
As stated in the PR description, I had done it. Then I messed it up and lost them. Thanks for catching it, fixing.

Reply:
Done.